### PR TITLE
[QUIDEM-80] Query recorder

### LINKED ISSUE #80: Query recorder
Add a component that can record queries to Quidem files and play them back.

Here is an example:
```
// The first recorder, in record mode, creates "foo.iq".
ConnectionFactory postgres;
Config config = Config.empty()
  .withFile("foo.iq")
  .withMode(RECORD)
  .withConnectionFactory(postgres);
Recorder recorder = Recorder.create(config);
recorder.executeQuery("select count(*) from emp", "empCount", result -> {
    assertThat(result.getMetaData().getColumnCount(), is(1));
    assertThat(result.next(), is(true));
    assertThat(result.getInt(1), is(13));
  });
recorder.close();

// The second recorder, in play mode, reads "foo.iq" created by the first step.
Config config2 = Config.empty()
  .withFile("foo.iq")
  .withMode(PLAY);
Recorder recorder2 = Recorder.create(config);
recorder.executeQuery("select count(*) from emp", result -> {
    assertThat(result.getMetaData().getColumnCount(), is(1));
    assertThat(result.next(), is(true));
    assertThat(result.getInt(1), is(13));
  });
recorder2.close();
```

Here is the `foo.iq` file that is generated in step 1 and read in step 2:
```
!use "postgres"

# empCount
select count(*) as c from emp;
c:int
======
13
!ok
```

### PR DESCRIPTION
**Pull request by julianhyde**
Fixes #80

### CODE REVIEW

File: src/main/java/net/hydromatic/quidem/record/JdbcUtils.java [hunk 1]
```diff
@@ -0,0 +1,1112 @@
...
+              fields = null;
+              return false;
+            }
+          }
+
+          @Override public boolean wasNull() {
+            return wasNull;
+          }
+
+          @Override public String getString(int columnIndex) {
```
Comment (jswett77) [inline src/main/java/net/hydromatic/quidem/record/JdbcUtils.java:218]: Having a way to represent an empty string different from null might be useful. Thinking of scenarios where we test grouping sets across edge cases, e.g. empty str, null in same query.

**Review by jswett77 — COMMENTED** [review]
The might be some useful testing scenarios where we want to disambiguate between null and empty string.

File: src/main/java/net/hydromatic/quidem/record/JdbcUtils.java (hunk 1 above)
Comment (julianhyde) [reply src/main/java/net/hydromatic/quidem/record/JdbcUtils.java:218]: Great point. I believe the encoding allows us to store both nulls and empty strings, but it's trickier with strings than with data types such as DATE and INTEGER. I will add a test. The following Postgres query seems pretty good:
```
# with t as (select "ENAME",
#   substring("ENAME", 1, mod("MGR",5) - 1) as s
#   from emp)
# select s, s is null, char_length(s)  from t;
 ENAME  |  s  | ?column? | char_length 
--------+-----+----------+-------------
 SMITH  | S   | f        |           1
 ALLEN  | AL  | f        |           2
 WARD   | WA  | f        |           2
 JONES  | JON | f        |           3
 MARTIN | MA  | f        |           2
 BLAKE  | BLA | f        |           3
 CLARK  | CLA | f        |           3
 SCOTT  |     | f        |           0
 KING   |     | t        |            
 TURNER | TU  | f        |           2
 ADAMS  | AD  | f        |           2
 JAMES  | JA  | f        |           2
 FORD   |     | f        |           0
 MILLER | M   | f        |           1
(14 rows)

```

Comment (julianhyde) [reply src/main/java/net/hydromatic/quidem/record/JdbcUtils.java:311]: Commit 7b1a50ffd091ce7e9ed394294dd03d549c164961 should fix this.

**Comment by julianhyde** [comment]
Fixed in https://github.com/hydromatic/quidem/commit/3b94aba4add60e93275bcb0e33c932aa0a417c7c.
